File: airtable.py
"""
Append chatter record to Airtable.
Field names: Name, Lastname, Replies, Status, Payment method, BTC address, wise name, Wise email, currency
Requires: AIRTABLE_API_KEY, AIRTABLE_BASE_ID, AIRTABLE_TABLE_NAME in .env
"""

import os
import json
import logging
import urllib.request
import urllib.error

logger = logging.getLogger(__name__)

AIRTABLE_API_URL = "https://api.airtable.com/v0"


def append_chatter(row: dict) -> None:
    """
    Create one record in the configured Airtable table.
    row: dict with keys: name, lastname, replies, status, payment_method,
         btc_address (optional), wise_name (optional), wise_email (optional), currency (optional)
    """
    api_key = os.getenv("AIRTABLE_API_KEY")
    base_id = os.getenv("AIRTABLE_BASE_ID")
    table_name = os.getenv("AIRTABLE_TABLE_NAME", "Chatters")

    if not api_key or not base_id:
        raise ValueError(
            "Missing Airtable config. Set AIRTABLE_API_KEY and AIRTABLE_BASE_ID in .env"
        )

    # Airtable field names must match exactly
    fields = {
        "Name": (row.get("name") or ""),
        "Lastname": (row.get("lastname") or ""),
        "Replies": (row.get("replies") or ""),
        "Status": (row.get("status") or ""),
        "Payment method": (row.get("payment_method") or ""),
        "BTC address": (row.get("btc_address") or ""),
        "wise name": (row.get("wise_name") or ""),
        "Wise email": (row.get("wise_email") or ""),
        "currency": (row.get("currency") or ""),
    }

    table_id_encoded = urllib.request.quote(table_name, safe="")
    url = f"{AIRTABLE_API_URL}/{base_id}/{table_id_encoded}"
    data = json.dumps({"fields": fields}).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        method="POST",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            if resp.status not in (200, 201):
                body = resp.read().decode()
                raise RuntimeError(f"Airtable API error {resp.status}: {body}")
    except urllib.error.HTTPError as e:
        body = e.read().decode() if e.fp else ""
        logger.exception("Airtable HTTP error %s: %s", e.code, body)
        raise RuntimeError(f"Airtable error {e.code}: {body}") from e
    logger.info("Appended chatter record to Airtable")

File: airtable.js
/**
 * Append one record to Airtable (passed applicants from OpenAI flow).
 * Requires: AIRTABLE_API_KEY, AIRTABLE_BASE_ID in .env
 * Optional: AIRTABLE_TABLE_NAME (default "Chatters")
 *
 * Fields: Name, Email, Phone, Telegram Username, Telegram User ID, Application notes, Status
 */

const https = require('https');

function request(path, body) {
  return new Promise((resolve, reject) => {
    const data = body ? JSON.stringify(body) : null;
    const req = https.request(
      {
        hostname: 'api.airtable.com',
        path: path.startsWith('/') ? path : `/${path}`,
        method: 'POST',
        headers: {
          Authorization: `Bearer ${process.env.AIRTABLE_API_KEY}`,
          'Content-Type': 'application/json',
        },
      },
      (res) => {
        let raw = '';
        res.on('data', (chunk) => { raw += chunk; });
        res.on('end', () => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(raw ? JSON.parse(raw) : {});
          } else {
            reject(new Error(`Airtable ${res.statusCode}: ${raw}`));
          }
        });
      }
    );
    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

async function appendApplicant(row) {
  const baseId = process.env.AIRTABLE_BASE_ID;
  const apiKey = process.env.AIRTABLE_API_KEY;
  const tableName = process.env.AIRTABLE_TABLE_NAME || 'Chatters';

  if (!apiKey || !baseId) {
    throw new Error('Missing Airtable config. Set AIRTABLE_API_KEY and AIRTABLE_BASE_ID in .env');
  }

  const tableEncoded = encodeURIComponent(tableName);
  const path = `/v0/${baseId}/${tableEncoded}`;

  const fields = {
    Name: row.fullName || '',
    Email: row.email || '',
    Phone: row.phone || '',
    'Telegram Username': row.telegramUsername || '',
    'Telegram User ID': row.telegramUserId || '',
    'Application notes': row.applicationNotes || '',
    Status: 'Passed',
  };

  await request(path, { fields });
}

module.exports = { appendApplicant };
